### PR TITLE
foliate: 2.6.4 -> 3.0.1

### LINKED ISSUE #260232: Update request: Update Foliate to the GTK4 branch.
- Package name: [foliate](https://github.com/johnfactotum/foliate)
- Latest released version: 2.6.4

- Current version on the unstable channel: 2.6.4
- Current version on the stable/release channel: 2.6.4

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

Seeing as the GTK3 branch (master) is pretty EOL and the app hasn't been updated for more than a year, I think it's a good idea to update to the latest commit of the GTK4 branch since that's the branch that being worked on.
**Notify maintainers**
@onny 

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/260232

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
GTK+libadw bump part of #247766

File: pkgs/applications/office/foliate/default.nix [hunk 1]
```diff
@@ -1,43 +1,70 @@
...
-  '';
+  nativeBuildInputs = [
+    gobject-introspection
+    meson
+    ninja
+    pkg-config
+    python3
+    wrapGAppsHook
+
+    cmake
```
Comment (jtojnar) [inline pkgs/applications/office/foliate/default.nix:48]: Are you sure `cmake` is needed? Meson supports it for finding dependencies but typically, `pkg-config` will be sufficient. See <https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/meson-cmake.md> for more info.

File: pkgs/applications/office/foliate/default.nix [hunk 2]
```diff
@@ -1,43 +1,70 @@
...
-    sed -i "1 a imports.package._findEffectiveEntryPointName = () => 'com.github.johnfactotum.Foliate';" $out/bin/.com.github.johnfactotum.Foliate-wrapped
-    ln -s $out/bin/com.github.johnfactotum.Foliate $out/bin/foliate
-  '';
+  nativeBuildInputs = [
+    gobject-introspection
+    meson
+    ninja
+    pkg-config
+    python3
+    wrapGAppsHook
```
Comment (jtojnar) [inline pkgs/applications/office/foliate/default.nix:46]: `wrapGAppsHook4` should be used with GTK 4.

File: pkgs/applications/office/foliate/default.nix [hunk 3]
```diff
@@ -1,43 +1,70 @@
...
+
+    cmake
+    libadwaita
+  ];
 
   buildInputs = [
+    desktop-file-utils
```
Comment (jtojnar) [inline pkgs/applications/office/foliate/default.nix:53]: Build tools should go to `nativeBuildInputs`, see https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md.

File: pkgs/applications/office/foliate/default.nix [hunk 4]
```diff
@@ -1,43 +1,70 @@
...
+  nativeBuildInputs = [
+    gobject-introspection
+    meson
+    ninja
+    pkg-config
+    python3
+    wrapGAppsHook
+
+    cmake
+    libadwaita
```
Comment (jtojnar) [inline pkgs/applications/office/foliate/default.nix:49]: Runtime dependencies should go to `buildInputs`, see https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md.

File: pkgs/applications/office/foliate/default.nix [hunk 5]
```diff
@@ -1,43 +1,70 @@
...
+    hash = "sha256-Nw6ABV0kqVeF6PGWSSoIXV8O0ymAfhUmdmF++BWSCtM=";
   };
 
-  nativeBuildInputs = [ meson ninja python3 wrapGAppsHook gobject-introspection ];
-
   postPatch = ''
     patchShebangs build-aux/meson/postinstall.py
 
     substituteInPlace src/main.js \
       --replace "'WebKit2': '4.0'" "'WebKit2': '4.1'"
```
Comment (jtojnar) [inline pkgs/applications/office/foliate/default.nix:37]: This should be no longer necessary.

File: pkgs/applications/office/foliate/default.nix [hunk 6]
```diff
@@ -1,43 +1,70 @@
...
     repo = pname;
     rev = version;
-    sha256 = "sha256-Pr2YA2MHXD4W7lyCxGAVLKyoZarZ8t92RSkWle3LNuc=";
+    hash = "sha256-Nw6ABV0kqVeF6PGWSSoIXV8O0ymAfhUmdmF++BWSCtM=";
   };
 
-  nativeBuildInputs = [ meson ninja python3 wrapGAppsHook gobject-introspection ];
-
   postPatch = ''
     patchShebangs build-aux/meson/postinstall.py
```
Comment (jtojnar) [inline pkgs/applications/office/foliate/default.nix:34]: Not used.

File: pkgs/applications/office/foliate/default.nix [hunk 7]
```diff
@@ -1,43 +1,70 @@
...
-, gobject-introspection, glib-networking }:
+{ stdenv
+, lib
+, fetchFromGitHub
+, meson
+, gettext
+, glib
+, gjs
+, ninja
+, python3
```
Comment (jtojnar) [inline pkgs/applications/office/foliate/default.nix:9]: Not needed.

File: pkgs/applications/office/foliate/default.nix [hunk 8]
```diff
@@ -1,43 +1,70 @@
...
-    desktop-file-utils
     gsettings-desktop-schemas
+    gtk4
+    webkitgtk_6_0
   ];
 
+  postFixup = ''
+    sed -i "1 a imports.package._findEffectiveEntryPointName = () => 'com.github.johnfactotum.Foliate';" $out/bin/.com.github.johnfactotum.Foliate-wrapped
+    ln -s $out/bin/com.github.johnfactotum.Foliate $out/bin/foliate
+  '';
```
Comment (johnfactotum) [inline pkgs/applications/office/foliate/default.nix:66]: These aren't needed anymore. In v3.0, the executable is called `foliate`, and it doesn't use GJS's package module, so setting `imports.package._findEffectiveEntryPointName` won't do anything.

**Pull request merged**